bestie/server/service.go

```diff
@@ -0,0 +1,133 @@
+)
+
+// List of event ID names defined by build_event_stream.proto.
+// These formal event names are mapped to Prometheus metric label names.
+var besEventIds = map[string]string{
+	"Unknown":                       "unknown", // should not normally see any of these
+	"Progress":                      "progress",
+	"Started":                       "started",
+	"CommandLine":                   "unstructured_command_line", // CommandLine replaced by UnstructuredCommandLine
+	"UnstructuredCommandLine":       "unstructured_command_line",
+	"StructuredCommandLine":         "structured_command_line",
+	"WorkspaceStatus":               "workspace_status",
+	"OptionsParsed":                 "options_parsed",
+	"Fetch":                         "fetch",
+	"Configuration":                 "configuration",
+	"TargetConfigured":              "target_configured",
+	"Pattern":                       "pattern",
+	"PatternSkipped":                "pattern_skipped",
+	"NamedSet":                      "named_set",
+	"TargetCompleted":               "target_completed",
+	"ActionCompleted":               "action_completed",
+	"UnconfiguredLabel":             "unconfigured_label",
+	"ConfiguredLabel":               "configured_label",
+	"TestResult":                    "test_result",
+	"TestSummary":                   "test_summary",
+	"TargetSummary":                 "target_summary",
+	"BuildFinished":                 "build_finished",
+	"BuildToolLogs":                 "build_tool_logs",
+	"BuildMetrics":                  "build_metrics",
+	"Workspace":                     "workspace",
+	"BuildMetadata":                 "build_metadata",
+	"ConvenienceSymlinksIdentified": "convenience_symlinks_identified",
+}
+
+// BES Prometheus metric definition.
+// Use this for untagged counters.
+type besPromMetric struct {
+	label string
+	ctr   prometheus.Counter
+}
+
+// BES Prometheus metric vector definition.
+// Use this for tagged counters.
+type besPromMetricVec struct {
+	ctr *prometheus.CounterVec
+}
+
+// BES Prometheus histogram definition.
+type besPromHistogram struct {
+	ctr prometheus.Histogram
+}
+
+// Number of BES messages seen by event type.const
+type besEvents struct {
+	bazelBuildsTotal besPromMetric
+	bazelEventsTotal besPromMetricVec
+}
+
+// Define BES Endpoint service counters being exposed through Prometheus.
+var (
+	promBazelBuildsTotal = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: "bes",
+			Subsystem: "srv",
+			Name:      "bazel_builds_total",
+			Help:      "Total number of Bazel builds seen",
+		})
+	promBazelEventsTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "bes",
+			Subsystem: "srv",
+			Name:      "bazel_events_total",
+			Help:      "Total observed events, tagged by event ID",
+		},
+		[]string{"id"},
+	)
+)
+
+// Service statistics.
+type serviceStats struct {
+	// General Build Event Stream stats.
+	besEvents
+}
+
+// Statistic values to report to Prometheus for this service.
```

Comment:
Why all the structs wrapping the metric types, only to declare the metrics as global? Why not use the global metrics directly?

e.g. https://github.com/enfabrica/enkit/blob/f61a0f0aa15fd7cf74fa0f94af10c43cc4499ab6/flextape/service/service.go#L21 defines some global metrics, that the service increments directly. This has extra wrapper structs, but global metrics are still used under the hood - what value do the wrapper structs provide?

Reply:
I was thinking that there might be other information related to each metric in addition to the actual Prometheus metric type itself. For example, the metric name used by Prometheus. I'll take another look at this to see if the wrapper provides any meaningful value-add.